cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -192,31 +194,30 @@ __global__ void __launch_bounds__(block_size, 1)
   auto map =
     map_type::device_view(chunk.dict_map_slots, chunk.dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
 
-  __shared__ size_type counter;
-  if (t == 0) counter = 0;
+  __shared__ cuda::atomic<size_type, cuda::thread_scope_device> counter;
+  using cuda::std::memory_order_relaxed;
+  if (t == 0) { counter.store(0, memory_order_relaxed); }
```

Comment:
This should use placement new, otherwise it's invoking `store` on an uninitialized object. Technically not a problem here, but it would cause a bug if `size_type` were larger than 8B (the internal lock bytes wouldn't be initialized).
```suggestion
  if (t == 0) { new (&counter) cuda::atomic<size_type, cuda::thread_scope_block>{0} }
```